modules/lib/zsh.nix

```diff
@@ -9,11 +9,20 @@ rec {
       ''"${v}"''
     else if builtins.isList v then
       "(${lib.concatStringsSep " " (map toZshValue v)})"
+    else if builtins.isAttrs v then
+      "(${
+        lib.concatStringsSep " "
+        (lib.mapAttrsToList (n: v: "[${lib.escapeShellArg n}]=${toZshValue v}")
+          v)
+      })"
     else
       ''"${toString v}"'';
 
   # Produces a Zsh shell like definition statement
-  define = n: v: "${n}=${toZshValue v}";
+  define = n: v:
+    "${lib.optionalString (builtins.isAttrs v) "typeset -A "}${n}=${
+      toZshValue v
+    }";
```

Comment:
Does it work with `exportAll`? If not, can we make it work with it?

Reply:
I don't see how as `export` is for environment variables and those are exclusively strings. And this deals with associative arrays, which are not strings.

Note, even though `typeset` has the `-x` option to "mark for automatic export to the environment of subsequently executed commands", it doesn't actually do anything for associative arrays. Likely because, without defining some serialization format, that doesn't make sense.

Example that demonstrates the difference between using `-x` for exporting to the environment:

```shell
$ typeset -x STR=regular-string-content
$ typeset -p STR # makes ZSH dump a command string that would duplicate this variable
export STR=regular-string-content
$ printenv STR
regular-string-content
$ echo $?
0
$ typeset -x -A ASSOC=([a]=mooh)
typeset -Ax ASSOC=( [a]=mooh )
$ printenv ASSOC
$ echo $?
1
```

Reply:
> those are exclusively strings

... Yeah, I had a brain fart there.

Thanks for the long answer though, TIL about `typeset -x`.